Fix: default sort option for discovery

buithaihai opened the pull request:
## References
Fix [#8837](https://github.com/DSpace/DSpace/issues/8837)

## Description

This PR introduces a default sort option for DiscoveryConfiguration in `discovery.xml`, which replaces default sorting (by `score,DESC`)

## Instructions for Reviewers

### List of changes in this PR:
1. Added `defaultSortField` to `DiscoverySortConfiguration`, whose value is configurable in `discovery.xml` like this:
(our use case bean `workspaceConfiguration`, refer to change list for actual changes):
```xml
[...]
    <bean id="workspaceConfiguration"
        class="org.dspace.discovery.configuration.DiscoveryConfiguration"
        scope="prototype">
        [...]
        <property name="searchSortConfiguration">
            <bean class="org.dspace.discovery.configuration.DiscoverySortConfiguration">
                <!--The default sort filter to use for the initial workspace loading-->
                <property name="defaultSortField" ref="sortDateAccessioned" />
        [...]
```
2. This leads to necessary changes in `dspace-api` and `webapp`, most notably `org.dspace.discovery.utils.DiscoverQueryBuilder#getDefaultSortDirection`, which now use `defaultSortField` if not null. In our case, DiscoverySearch for workspace configuration will sort by `sortDateAccessioned,DESC` from now on.
3. To accomodate **dspace-angular**, which sort using the first option in `<property name="sortFields"/>` in `discovery.xml`, we add reference to `<bean id="sortDateAccessioned"/>` to the list value
```xml
                <property name="sortFields">
                    <list>
                        <ref bean="sortDateAccessioned" /> <!--addition-->
                        <ref bean="sortScore" />
                        <ref bean="sortTitle" />
                        <ref bean="sortDateIssued" />
                    </list>
                </property>
```

### For Testing:
Added 2 test cases:
 - `org.dspace.app.rest.utils.RestDiscoverQueryBuilderTest#testSortByDefaultSortField`: verify that DiscoverySortConfiguration.defaultSortField is set and can be used for sorting
 - `org.dspace.discovery.DiscoveryIT#searchWithDefaultSortServiceTest`: verify that `DiscoverResult.indexableObjects` of a `DiscoverQuery` in `workspace` configuration is sorted by configured `defaultSortField`

## Checklist
- [x] My PR is small in size (e.g. less than 1,000 lines of code, not including comments & integration tests). Exceptions may be made if previously agreed upon.
- [x] My PR passes Checkstyle validation based on the [Code Style Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Style+Guide).
- [x] My PR includes Javadoc for _all new (or modified) public methods and classes_. It also includes Javadoc for large or complex private methods.
- [x] My PR passes all tests and includes new/updated Unit or Integration Tests based on the [Code Testing Guide](https://wiki.lyrasis.org/display/DSPACE/Code+Testing+Guide).
- [x] If my PR includes new libraries/dependencies (in any `pom.xml`), I've made sure their licenses align with the [DSpace BSD License](https://github.com/DSpace/DSpace/blob/main/LICENSE) based on the [Licensing of Contributions](https://wiki.lyrasis.org/display/DSPACE/Code+Contribution+Guidelines#CodeContributionGuidelines-LicensingofContributions) documentation.
- [x] If my PR modifies REST API endpoints, I've opened a separate [REST Contract](https://github.com/DSpace/RestContract/blob/main/README.md) PR related to this change.
- [x] If my PR includes new configurations, I've provided basic technical documentation in the PR itself.
- [x] If my PR fixes an issue ticket, I've [linked them together](https://docs.github.com/en/issues/tracking-your-work-with-issues/linking-a-pull-request-to-an-issue).

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1204642791317008

paulo-graca reviewed (commented):
I've tested and it worked kind I was expecting (I had to make some changes). First I used 
`<property name="defaultSortField" ref="sortLastModified" />`

But I understood to have it working properly, solr sorting will look for something like:
`&sort=lastModified_dt+desc&`
and there is no lastModified_dt field. We need to created it in solr schema...

I've added a copyField for lastModified_dt in search schema.xml:
```xml
    <field name="lastModified" type="date" indexed="true" stored="true" default="NOW" multiValued="false" omitNorms="true" />

    <copyField source="lastModified" dest="lastModified_dt" />

    <!-- used to filter out items that are older versions of another item -->
```
after reindexing all the content, my dspace worked as expected.

paulo-graca commented on dspace/config/spring/api/discovery.xml:
```diff
@@ -865,8 +865,11 @@
         <property name="searchSortConfiguration">
             <bean class="org.dspace.discovery.configuration.DiscoverySortConfiguration">
+                <!--The default sort filter to use for the initial workspace loading-->
+                <property name="defaultSortField" ref="sortDateAccessioned" />
```
@buithaihai thank you so much for your contribution. It would help a lot many users. 
Can you consider `lastModified` as the default sort field and also proposing `copyField` an addition to search Solr schema?

paulo-graca commented:
@buithaihai thank you so much for your contribution. It would help a lot many users. 
Can you consider `lastModified` as the default sort field and also proposing `copyField` an addition to search Solr schema?

buithaihai committed:
Fix: default sort option (lastModified) for discovery

buithaihai commented:
> Can you consider `lastModified` as the default sort field and also proposing `copyField` an addition to search Solr schema?

Yes. Testing indicates that this is good and produced no problem. I have modified `discovery.xml` code so that it works as you describe:
```xml
                <!--The default sort filter to use for the initial workspace loading-->
                <property name="defaultSortField" ref="sortDateAccessioned" />
                <property name="sortFields">
                    <list>
                        <ref bean="sortDateAccessioned" />
                        <ref bean="sortScore" />
                        <ref bean="sortTitle" />
                        <ref bean="sortDateIssued" />
                    </list>
                </property>
```
P/s: I rebased then force-pushed my code to make it up-to-date on `origin/DS-8837`, so if you want to verify it again I suggest you delete the local branch and then checkout again.

paulo-graca reviewed (commented):
Thank you again @buithaihai for addressing the solr change part!
I've some minor change suggestions.

buithaihai committed:
Minor Tweaks

buithaihai commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/DiscoveryRestControllerIT.java:
```diff
@@ -1286,8 +1286,10 @@ public void discoverSearchByFieldNotConfiguredTest() throws Exception {
+        // which will return status 400 instead of 422 if left unchanged
         getClient().perform(get("/api/discover/search/objects")
-                   .param("sort", "dc.date.accessioned, ASC")
+                   .param("sort", "person.familyName, ASC")
```
True that, I changed it back

buithaihai commented:
I have made the requested changes. Somehow the change to `sortLastModified` failed. It should have been as you said.

paulo-graca commented:
Thank you again @buithaihai this work is very helpful for users!
I will test it again today.

paulo-graca reviewed (approved):
I've tested. It's working as expected.

@tdonohue this PR needs to change search solr schema in order to have sortLastModified sort working as expected. I noticed that discovery config: `sortLastModified` didn't worked because Discovery will add a prefix "_dt" for the solr query an that specific field (`lastModified_dt`) didn't exist.

tdonohue reviewed (changes requested):
@buithaihai : Thanks for this fix!  Overall, I've verified this works well, so I'm nearly a +1.

But, I have a few questions/comments in the tests you've provided.  I think these tests can be improved, and might accidentally be returning false positives in some scenarios (which is never a good thing).  If I'm misunderstanding the code though, let me know.

Overall, I like this improvement.  I will note for other testers that this bug fix *does* require a change the Solr schema.xml for the "search" core.  That means to fix this bug you *must* reindex (`./dspace index-discovery -b`).

If you can fix the tests or respond to my comments within the next week, then I think we can get this merged in time for 7.6

buithaihai commented on dspace-server-webapp/src/test/java/org/dspace/app/rest/utils/RestDiscoverQueryBuilderTest.java:
```diff
@@ -167,6 +169,19 @@ public void testSortByScore() throws Exception {
 
+    @Test
+    public void testSortByDefaultSortField() throws Exception {
+        page = PageRequest.of(2, 10, Sort.Direction.DESC, "dc.date.accessioned");
```
You are right, I somehow failed to figure it out. Changed these lines per suggestion

buithaihai committed:
Tweaks fo test cases.

buithaihai commented on dspace-api/src/test/java/org/dspace/discovery/DiscoveryIT.java:
```diff
@@ -731,6 +737,68 @@ public void iteratorSearchServiceTest() throws SearchServiceException {
+
+        DiscoverResult result = searchService.search(context, discoverQuery);
+
+        if (defaultSortField.getMetadataField().equals("dc_date_accessioned")) {
```
Noted. I commented out the codes that test against `dc.date.accesioned`. It was a bit of a pain to figure it out and I wanted to keep it in case we have to check against it again.

buithaihai commented on dspace-api/src/test/java/org/dspace/discovery/DiscoveryIT.java:
```diff
@@ -731,6 +737,68 @@ public void iteratorSearchServiceTest() throws SearchServiceException {
+
+        // Populate the testing objects: create items in eperson's workspace and perform search in it
+        int numberItems = 10;
+        context.turnOffAuthorisationSystem();
```
It is done, please check the new commit.

paulo-graca committed:
Merge branch 'main' into DS-8837

paulo-graca committed:
Merge branch 'main' into DS-8837

tdonohue reviewed (approved):
👍 Thanks @buithaihai !  This looks good to me now & it fixes the sorting issue in MyDSpace.

alanorth commented:
Hi @buithaihai, from this pull request I see it should be possible to set the `defaultSortField` for a bean.

I am using DSpace 8.3 and would like the collection search pages to default to `sortDateAccessionedDesc`. I tried to set it for the `collectionConfiguration` bean as follows:

```xml
    <bean id="collectionConfiguration" class="org.dspace.discovery.configuration.DiscoveryConfiguration"
          scope="prototype" parent="defaultConfiguration">
        <!--The sort filters for the discovery search-->
        <property name="searchSortConfiguration">
            <bean class="org.dspace.discovery.configuration.DiscoverySortConfiguration">
                <property name="defaultSortField" ref="sortDateAccessionedDesc" />
                <property name="sortFields">
                    <list>
                        <ref bean="sortDateAccessioned"/>
                        <ref bean="sortDateAccessionedDesc"/>
                        <ref bean="sortScore"/>
                        <ref bean="sortTitle"/>
                        <ref bean="sortDateIssued"/>
                    </list>
                </property>
            </bean>
        </property>
    </bean>
```

But this does not work. The only thing that works is setting `sortDateAccessionedDesc` first in the list of sort fields.

*Edit:* I see this frontend issue mentioning the exact problem: https://github.com/DSpace/dspace-angular/issues/2524
